src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -3603,6 +3605,15 @@ public override BoundNode VisitConversion(BoundConversion node)
                 reportTopLevelWarnings: fromExplicitCast,
                 reportRemainingWarnings: true,
                 trackMembers: true);
+
+            if ((node.ConversionKind == ConversionKind.ExplicitUserDefined ||
+                 node.ConversionKind == ConversionKind.ImplicitUserDefined) &&
+                 !(conversion.Method is null))
+            {
+                VisitOperandsHonoringAnnotations(conversion.Method, ImmutableArray.Create(operand));
```

Comment:
>VisitOperandsHonoringAnnotations [](start = 16, length = 32)

Was there a reason this could not be pushed into `ApplyConversion`?
Also, there are tuple scenarios worth testing: A converts to B with user-defined conversion, then `(B, B) t = (a, a);` and variants.

I'm wondering whether we should ignore annotations on such user-defined conversions though. It's not obvious what is the benefit...

Reply:
There may also be a problem with not saving/restoring the state here, given that we visit the operand twice. Maybe a scenario like `(x.String() + (x = null))--` with a user-defined `-`.

---
In reply to: [270163123](https://github.com/dotnet/roslyn/pull/34556#discussion_r270163123) [](ancestors = 270163123)

Reply:
I originally prototyped it in ApplyConversion, but was trying to make the change smaller: we'd need to handle a potentially split state in other places if we have it in apply conversion. 

I'm not sure how the tuple conversions would be used? you could obviously annotate them, but are there scenarios where they would actually have any effect?

---
In reply to: [270163123](https://github.com/dotnet/roslyn/pull/34556#discussion_r270163123) [](ancestors = 270163123)